Fix: set $submitted on child forms when form is submitted

Narretz opened the pull request:
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
Fix


**What is the current behavior? (You can also link to an open issue here)**
When submitting a form, $submitted is not set on child forms


**Does this PR introduce a breaking change?**
I don't think so.
No one in #10071 said that his might be breaking.
AngularJS doesn't have nested form isolation, and ngForm is only for grouping sub-forms. I think that it's expected that a submission event will affect all sub-forms.
Probably: https://github.com/angular/angular.js/pull/11023#issuecomment-143162373

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


Follow up to https://github.com/angular/angular.js/pull/11023

petebacondarwin commented on src/ng/directive/form.js:
```diff
@@ -268,12 +268,22 @@ FormController.prototype = {
-    this.$$animate.addClass(this.$$element, SUBMITTED_CLASS);
-    this.$submitted = true;
-    this.$$parentForm.$setSubmitted();
+  $setSubmitted: function(onlySetOnChildren) {
```
The `onlySetOnChildren` param name is misleading as it might imply that we don't set it on `this` :-)
How about something about not setting it on the parent?

petebacondarwin replied:
Is this param even necessary? I don't see it used anywhere?
Perhaps it would be best to leave it out for this PR?

Narretz replied:
It's used in line 278 - we need it because if we always set it on both parent and children we can go into an infinite loop. 
You're right with the parameter name.

petebacondarwin replied:
oh right!
+1 on the new param name

mgol replied:
This is another boolean trap in a public API, isn't it? :/

mgol replied:
We have quite a lot of boolean parameters in our providers config methods but then they're not *so* bad as the convention is shared across the library and in all cases it just enables or disables the setting indicated by the method name.

Here there's no obvious relation between the parameter and the method name; it's impossible to know what's happening just by looking at the code. Naming the variable properly only helps us. Note, though, that negative parameters are confusing to people as when you set it to `false` you have double negation and it's been scientifically proven that it's harder for people to process.

If this is going to extent the public API, I think we should use an object with a parameter so that it's named (and then rather `setOnParent` than `doNotSetOnParent`). It's way easier to do it in ES6 but it's still possible in ES5.

petebacondarwin replied:
The alternative is to create a new private function. Perhaps something like this:

```
$setSubmitted: function() {
  var rootForm = this;
  while(rootForm.$$parentForm) { rootForm = rootForm.$$parentForm; }
  rootForm.$$setSubmitted();
}

$$setSubmitted() {
  this.$$animate.addClass(this.$$element, SUBMITTED_CLASS);
  this.$submitted = true;
  forEach(this.$$controls, function(control) {
    if (control.$$setSubmitted && !control.$submitted) {
      control.$setSubmitted(false);
    }
  });
}
```

Narretz commented:
@petebacondarwin wdyt about the breaking-changeness of the PR? @gkalpak had some reservations in the previous PR: https://github.com/angular/angular.js/pull/11023#issuecomment-143162373

petebacondarwin commented:
Regarding the BC - I guess that it is a breaking change. And this also raises the importance of being able to "detach" or "isolate" subforms from their parent. In which case you could block the child/parent form from being submitted if it was not "attached" to the form on which the setSubmit was called.

Narretz commented:
Ok, then let's put into 1.7.0 and rethink the API.

aeslinger0 commented:
I'm not sure why this needs to be a breaking change. I'm probably missing something. If the goal is to set child forms to submitted when the parent form is submitted, couldn't you add an attribute to ngForm such as `inherit-submit="true"` and use that as a flag to conditionally apply the new functionality? Since it'd be an optional attribute that results in the current behavior when omitted, I don't think it'd be breaking. Plus, it'd give you the ability to control form-by-form which subforms inherit the submit action.

petebacondarwin commented:
@aeslinger0 that is an interesting idea. What do you think @Narretz ?

Narretz commented:
Hm, I have to think about it, but in general the form API isn't very consistent as is, and I don't think an option that affects a single part of the API isn't very elegant.

Narretz commented:
I've implemented @petebacondarwin 's approach to setting submitted.
Wrt the need for having a way to detach the form from its parent, here are a few notes on the consequences that would follow https://github.com/angular/angular.js/pull/10193#r66080985

Narretz commented:
Here's a re-cap on what I think would be needed to support "detached" forms, i.e. a form that does not inherit or propagate state from / to its parent form.

Requirements:
1. A way to detach / re-attach a form. This can be done with $removeControl on linking, and could be wrapped in a directive that watches an expression to toggle this behavior. (When detaching / re-attaching later, the state would only be propagated after the form changes again)

2. To specify / implement the behavior of a detached form with regard to form submission.

Problems:
2a) the submit rules for forms (see https://code.angularjs.org/1.6.7/docs/api/ng/directive/form#submitting-a-form-and-preventing-the-default-action) also only apply to form elements. i.e. if you have ngForm with only one text input, pressing enter in the text input should trigger a submit event. 

2b) only the form element fires a "submit" event when you press a submit button etc., ngForm does not. This means, if you want to submit detached forms individually, you would have to wrap forms in ngForm. That however would mean, you cannot submit the whole form. 
Also, if you have form -> ngForm, and you press a submit button in the detached ngForm, the parent form will still receive the submit event.

Solution:
Imo can both 1) and 2) be implemented in user land - in fact, there are already libs that do 1) (although not perfect). For 2), a) is also already partly available (https://github.com/MarkPieszak/Angular1-ngForm-ngSubmit-fixes).
b) is more difficult, but should be possible by firing synthetic submit event on ngForm and overriding the form
submission handling with a function that checks if the submit event was triggered by an element that sits inside a detached ngForm (in which case it would not be fired on this form).

We should also update the docs with the info that in core ngForm, when used standalone, does not support form-specific submit handling.

Finally, I would introduce this PR as the default, without an option - if you have nested forms that are not detached, then I find it reasonable that the parent form sets $submitted on child forms.

gkalpak commented:
I still don't like the idea of detached forms fwiw :grin:

Narretz commented:
That's why I would move this whole topic to user-land. I agree that the semantics are unclear, it can become very complex very quickly, we don't know what kind of behavior devs actually want.
From my perspective, we only need to decide if this PR is breaking or not, or if we should make it configurable.

petebacondarwin reviewed (approved):
This PR looks good to me as it stands. Let's punt on detached forms for the moment.

Narretz commented:
Ok, so then I will squash the commits, add a BC notice and merge only into master.

gkalpak commented on src/ng/directive/form.js:
```diff
@@ -274,12 +275,25 @@ FormController.prototype = {
     this.$submitted = true;
-    this.$$parentForm.$setSubmitted();
+    forEach(this.$$controls, function(control) {
+      if (control.$$setSubmitted && !control.$submitted) {
```
Why the `!control.$submitted` check? Under what circumstances can a sub-form be `$submitted` (while the parent form is not) and does that guarantee that all sub-forms of the sub-form are also `$submitted`?

Narretz replied:
Technically, the form could have been removed with "removeControl", then have been changed to submitted via "setSubmitted" and re-attached to the parent form with "addControl". Unlikely, but possible. I don't think it's worth optimzing this, though. I think I took the implementation from this comment: https://github.com/angular/angular.js/pull/15778#discussion_r107630279

Narretz committed:
fix(form): set $submitted to true on child forms when parent is submitted

Closes #10071

BREAKING CHANGE:

Forms will now set $submitted on child forms when they are submitted.
For example:
```
<form name="parentform" ng-submit="$ctrl.submit()">
  <ng-form name="childform">
    <input type="text" name="input" ng-model="my.model" />
  </ng-form>
  <input type="submit" />
</form>

Submitting this form will set $submitted on "parentform" and "childform".
Previously, it was only set on "parentform".

This change was introduced because mixing form and ngForm does not create
logically separate forms, but rather something like input groups.
Therefore, child forms should inherit the submission state from their parent form.

Narretz committed:
docs(ngForm): clarify usage and limitations
